File: Easy_Leetcode_Problems/Leetcode_question_3090.java
// We use the Sliding Window technique.
//
// 1. Use two pointers: left and right.
//
// 2. Move right through the string.
//
// 3. Store the frequency of each character
//    in an integer array.
//
// 4. Each character can appear at most 2 times.
//
// 5. If a character appears more than 2 times,
//    move left forward until its frequency
//    becomes at most 2.
//
// 6. For every valid window, calculate:
//
//       right - left + 1
//
// 7. Store the maximum length.

// Step 1: Create an integer array of size 26
//         to store character frequencies.
//
// Step 2: Initialize:
//
//       left = 0
//       ans = 0
//
// Step 3: Traverse the string using right.
//
// Step 4: Increase frequency of s.charAt(right).
//
// Step 5: While the frequency of the current
//         character is greater than 2:
//
//       decrease frequency of s.charAt(left)
//       move left forward
//
// Step 6: Calculate current window length:
//
//       right - left + 1
//
// Step 7: Update ans.
//
// Step 8: Return ans.

// --------- Dry Run ------------

// Input:
// s = "aaaa"
//
// Initially:
//
// left = 0
// ans = 0
//
// --------------------------------
//
// right = 0
// window = "a"
// count[a] = 1
// length = 1
// ans = 1
//
// --------------------------------
//
// right = 1
// window = "aa"
// count[a] = 2
// length = 2
// ans = 2
//
// --------------------------------
//
// right = 2
// count[a] = 3 ❌
//
// 'a' occurs more than 2 times.
//
// Move left:
//
// remove s[0] = 'a'
// count[a] = 2
// left = 1
//
// window = "aa"
// length = 2
//
// ans = 2
//
// --------------------------------
//
// right = 3
// count[a] = 3 ❌
//
// Remove s[1]:
//
// count[a] = 2
// left = 2
//
// window = "aa"
// length = 2
//
// --------------------------------
//
// Final Answer:
//
// 2

class Solution {
    public int maximumLengthSubstring(String s) {

        int[] count = new int[26];

        int left = 0;
        int ans = 0;

        for (int right = 0; right < s.length(); right++) {

            char ch = s.charAt(right);
            count[ch - 'a']++;

            while (count[ch - 'a'] > 2) {

                count[s.charAt(left) - 'a']--;
                left++;
            }

            ans = Math.max(ans, right - left + 1);
        }

        return ans;
    }
}

// Interview Explanation--

// I used the Sliding Window technique.
//
// I maintained two pointers, left and right,
// to represent the current substring.
//
// I used an integer array of size 26 to store
// the frequency of each lowercase character.
//
// I expanded the window using the right pointer.
//
// If any character appeared more than two times,
// I moved the left pointer forward and decreased
// its frequency until the window became valid.
//
// For every valid window, I calculated its length
// and updated the maximum length.
//
// Time Complexity: O(n)
//
// Space Complexity: O(1)
//
// The space is O(1) because there are only
// 26 lowercase English letters.

//  The key idea to remember for interview:

// Expand → Check frequency → Shrink → Update maximum

File: Easy_Leetcode_Problems/Leetcode_question_3090.py
# We use the Sliding Window technique.
#
# 1. Use two pointers: left and right.
#
# 2. Move right through the string and add each
#    character to the frequency dictionary.
#
# 3. Every character can occur at most 2 times.
#
# 4. If the current character occurs more than
#    2 times, move left forward until its count
#    becomes at most 2.
#
# 5. For every valid window, calculate:
#
#       length = right - left + 1
#
# 6. Keep the maximum length in ans.

# Step 1: Create an empty frequency dictionary.
#
# Step 2: Set:
#
#       left = 0
#       ans = 0
#
# Step 3: Traverse the string using right.
#
# Step 4: Increase the frequency of s[right].
#
# Step 5: While frequency of s[right] > 2:
#
#       decrease count of s[left]
#       move left forward
#
# Step 6: Calculate the current window length.
#
# Step 7: Update ans with the maximum length.
#
# Step 8: Return ans.

##----------- Dry Run ---------##

# Input:
# s = "bcbbbcba"
#
# We need a substring where every character
# appears at most 2 times.
#
# Initially:
#
# left = 0
# ans = 0
#
# ------------------------------------------------
#
# right = 0
# window = "b"
# b = 1
# length = 1
# ans = 1
#
# ------------------------------------------------
#
# right = 1
# window = "bc"
# b = 1, c = 1
# length = 2
# ans = 2
#
# ------------------------------------------------
#
# right = 2
# window = "bcb"
# b = 2, c = 1
# length = 3
# ans = 3
#
# ------------------------------------------------
#
# right = 3
# window = "bcbb"
# b = 3 ❌
#
# b occurs more than 2 times.
# Move left:
#
# remove s[0] = 'b'
# b = 2
# left = 1
#
# Valid window = "cbb"
# length = 3
#
# ans = 3
#
# ------------------------------------------------
#
# right = 4
# window = "cbbb"
# b = 3 ❌
#
# Move left:
#
# remove 'c'
# c = 0
# left = 2
#
# remove 'b'
# b = 2
# left = 3
#
# Valid window = "bb"
# length = 2
#
# ans = 3
#
# ------------------------------------------------
#
# The remaining characters are processed
# in the same way.
#
# The maximum valid substring length is 4.
#
# Final Answer:
# 4

class Solution:
    def maximumLengthSubstring(self, s: str) -> int:

        count = {}
        left = 0
        ans = 0

        for right in range(len(s)):

            count[s[right]] = count.get(s[right], 0) + 1

            while count[s[right]] > 2:
                count[s[left]] -= 1
                left += 1

            ans = max(ans, right - left + 1)

        return ans

# I used the Sliding Window technique to solve
# this problem efficiently.
#
# I maintained two pointers, left and right,
# to represent the current substring.
#
# I also used a dictionary to store the frequency
# of each character.
#
# I expanded the window using the right pointer.
#
# If any character appeared more than two times,
# I moved the left pointer forward and decreased
# the corresponding frequencies until the window
# became valid again.
#
# For every valid window, I calculated its length
# and stored the maximum length.
#
# Time Complexity: O(n)
#
# Space Complexity: O(1)
#
# Because there are only 26 lowercase English
# letters, the frequency dictionary contains
# at most 26 characters.

